src/commonMain/kotlin/fr/acinq/lightning/wire/OfferTypes.kt

```diff
@@ -0,0 +1,1003 @@
+             * @param features    invoice features.
+             * @param chain       chain on which the offer is valid.
+             */
+            operator fun invoke(
+                amount_opt: MilliSatoshi?,
+                description: String,
+                nodeId: PublicKey,
+                features: Features,
+                chain: BlockHash,
+                additionalTlvs: Set<OfferTlv> = setOf(),
+                customTlvs: Set<GenericTlv> = setOf()
+            ): Offer {
+                val tlvs: Set<OfferTlv> = setOfNotNull(
+                    if (chain != Block.LivenetGenesisBlock.hash) OfferChains(listOf(chain)) else null,
+                    amount_opt?.let { OfferAmount(it) },
+                    OfferDescription(description),
+                    if (features != Features.empty) OfferFeatures(features) else null,
+                    OfferNodeId(nodeId),
+                ) + additionalTlvs
+                return Offer(TlvStream(tlvs, customTlvs))
+            }
+
+            fun validate(records: TlvStream<OfferTlv>): Either<InvalidTlvPayload, Offer> {
+                if (records.get<OfferDescription>() == null) return Left(MissingRequiredTlv(10L))
+                if (records.get<OfferNodeId>() == null) return Left(MissingRequiredTlv(22L))
+                if (records.unknown.any { it.tag >= 80 })
+                    return Left(ForbiddenTlv(records.unknown.find{it.tag >= 80}!!.tag))
+                return Right(Offer(records))
+            }
+
+            val tlvSerializer = TlvStreamSerializer(
+                false, @Suppress("UNCHECKED_CAST") mapOf(
+                    OfferChains.tag to OfferChains as TlvValueReader<OfferTlv>,
+                    OfferMetadata.tag to OfferMetadata as TlvValueReader<OfferTlv>,
+                    OfferCurrency.tag to OfferCurrency as TlvValueReader<OfferTlv>,
+                    OfferAmount.tag to OfferAmount as TlvValueReader<OfferTlv>,
+                    OfferDescription.tag to OfferDescription as TlvValueReader<OfferTlv>,
+                    OfferFeatures.tag to OfferFeatures as TlvValueReader<OfferTlv>,
+                    OfferAbsoluteExpiry.tag to OfferAbsoluteExpiry as TlvValueReader<OfferTlv>,
+                    OfferPaths.tag to OfferPaths as TlvValueReader<OfferTlv>,
+                    OfferIssuer.tag to OfferIssuer as TlvValueReader<OfferTlv>,
+                    OfferQuantityMax.tag to OfferQuantityMax as TlvValueReader<OfferTlv>,
+                    OfferNodeId.tag to OfferNodeId as TlvValueReader<OfferTlv>,
+                )
+            )
+
+            fun decode(s: String): Offer {
+                val (prefix, encoded, encoding) = Bech32.decodeBytes(s.lowercase(), true)
+                require(prefix == hrp)
+                require(encoding == Bech32.Encoding.Beck32WithoutChecksum)
```

Comment:
We shouldn't throw exceptions here, it will crash the app when scanning something that isn't an offer (we've seen issues like this when parsing invalid Bolt 11 invoices that threw exceptions and fixed it since). The `decode` functions (all of them) should return a `Try`.